src/core/namespace.c

```diff
@@ -1851,54 +1883,93 @@ int setup_netns(const int netns_storage_socket[static 2]) {
          *
          * It's a bit crazy, but hey, works great! */
 
-        if (lockf(netns_storage_socket[0], F_LOCK, 0) < 0)
-                return -errno;
+        r = lockfp(netns_storage_socket[0], &storage_socket0_lock);
+        if (r < 0)
+                return r;
 
         netns = receive_one_fd(netns_storage_socket[0], MSG_DONTWAIT);
         if (netns == -EAGAIN) {
                 /* Nothing stored yet, so let's create a new namespace. */
 
-                if (unshare(CLONE_NEWNET) < 0) {
-                        r = -errno;
-                        goto fail;
-                }
+                if (unshare(CLONE_NEWNET) < 0)
+                        return -errno;
 
                 (void) loopback_setup();
 
                 netns = open("/proc/self/ns/net", O_RDONLY|O_CLOEXEC|O_NOCTTY);
-                if (netns < 0) {
-                        r = -errno;
-                        goto fail;
-                }
+                if (netns < 0)
+                        return -errno;
 
                 r = 1;
 
-        } else if (netns < 0) {
-                r = netns;
-                goto fail;
+        } else if (netns < 0)
+                return netns;
 
-        } else {
+        else {
                 /* Yay, found something, so let's join the namespace */
-                if (setns(netns, CLONE_NEWNET) < 0) {
-                        r = -errno;
-                        goto fail;
-                }
+                if (setns(netns, CLONE_NEWNET) < 0)
+                        return -errno;
 
                 r = 0;
         }
 
         q = send_one_fd(netns_storage_socket[1], netns, MSG_DONTWAIT);
-        if (q < 0) {
-                r = q;
-                goto fail;
-        }
+        if (q < 0)
+                return q;
 
-fail:
-        (void) lockf(netns_storage_socket[0], F_ULOCK, 0);
         return r;
 }
 
+static int netns_create(const char *path, bool create) {
+        _cleanup_close_ int old_netns = -1, netns = -1;
+        int r;
+
+        if (create) {
+                r = mkdir_parents(path, 0755);
+                if (r < 0)
+                        return r;
+
+                r = touch(path);
+                if (r < 0)
+                        return r;
+        }
+
+        /* Save the current netns. */
+        old_netns = open("/proc/self/ns/net", O_RDONLY | O_CLOEXEC);
+        if (old_netns < 0)
+                return -errno;
+
+        /* Create a new netns */
+        if (unshare(CLONE_NEWNET) < 0)
+                return -errno;
+
+        (void) loopback_setup();
+
+        /* Mount the new netns onto the path. */
+        if (mount("/proc/self/ns/net", path, "none", MS_BIND, NULL) < 0) {
```

Comment:
Do we ever clean this up somewhere? Won't we end up with a namespace leak if we keep the file bind-mounted somewhere?

Reply:
We do not know when the file is not used anymore. So, I think we must 'leak' the file. (And can be manually removed by `networkctl remove-netns` command.)

Reply:
Maybe there's a way to keep track of whether there are any services running in the network namespace (or just a namespace in general to be a little more generic)? As soon as the count reaches zero, we unmount the file. Of course, that does not keep track of any processes not spawned by systemd but that seems like a tradeoff we could make (and explicitly document).

The idea is that we only need to have the file mounted if there is at least one other service running in the namespace. If the last service exits, we can safely remove the mount and create the mount again once another service starts with the same namespace path.

The only thing I'm not sure about is how hard it is to add the machinery required to keep track of which services are running in which persistent namespaces.

On the other hand, leaking the file might not be a huge problem. Although I'm a bit worried that if many unique instances of a template service are spawned and they use the template parameter in the network namespace path that we get a lot of namespace mounts lying around. I have a feeling someone would hit some kind of limit sooner or later and get a rather obscure error if we don't clean the mounts ourselves.

Maybe @poettering knows more?

Reply:
I noticed that removing the netns file causes backward compatibility issue. Previously, when a service which has `NetworkNamespacePath=` is stopped, the netns file still exists, and can be used any later commands or invoked services. Such behavior must be preserved. So, even if pid1 creates the file, the file must be 'leak'ed.

Reply:
Well yes, but that's only because the netns file was created explicitly by the user beforehand no? It makes sense then that they can rely on the file being there after the service stops.

However, if the file doesn't exist, we currently hard-fail so I think that means we have an avenue for adding new behaviour any way we want to without having to worry about backwards compatibility. In other words, there's no way users can depend on netns files created by systemd being there after the service stops since at the moment systemd never creates any netns files in the first place.

If the user currently creates the netns file, of course he's responsible for cleaning it up afterwards. However, if we now start creating the netns files ourselves, I think that makes us responsible for cleaning them up as well.

Reply:
I wonder if /var/run/netns mountpoint could be handled by systemd .mount & service lifetime/dependencies.

Reply:
I'd probably not actively use .mount unit for this kind of mount point. It's not really a file system after all, but more a weird API of the kernel to make namespaces persistent...

I'd just add NetworkNamespaceMode=join|create|sticky instead, as suggested elsewhere.